File: deploy_contract.py
#!/usr/bin/env python3
"""Deploy StateBloater contract using web3.py"""

import json
from web3 import Web3

# Connect to Besu
w3 = Web3(Web3.HTTPProvider('http://localhost:8545'))

print('Deploying StateBloater contract...\n')

# Check connection
if not w3.is_connected():
    print('❌ Could not connect to Besu')
    print('Start Besu with: ./scripts/start_besu_baseline.sh')
    exit(1)

print(f'✓ Connected to Besu')
print(f'  Current block: {w3.eth.block_number}')

# Load contract
with open('StateBloater.json', 'r') as f:
    artifact = json.load(f)

# Account setup
private_key = '0x8f2a55949038a9610f502c24114d051185071191bc20b60811a2d7fba4513689'
account = w3.eth.account.from_key(private_key)

print(f'✓ Using deployer: {account.address}')

# Create contract object
Contract = w3.eth.contract(abi=artifact['abi'], bytecode=artifact['bytecode'])

print('\n🚀 Deploying contract...')

# Build transaction
tx = Contract.constructor().build_transaction({
    'from': account.address,
    'nonce': w3.eth.get_transaction_count(account.address),
    'gas': 8000000,
    'gasPrice': 0,
    'chainId': 1337
})

# Sign and send
signed_tx = account.sign_transaction(tx)
tx_hash = w3.eth.send_raw_transaction(signed_tx.raw_transaction)

print(f'✓ Transaction hash: {tx_hash.hex()}')
print('  Waiting for confirmation...')

# Wait for receipt
receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=60)
contract_address = receipt.contractAddress

print('\n' + '=' * 60)
print('✅ DEPLOYMENT SUCCESSFUL')
print('=' * 60)
print(f'Contract Address: {contract_address}')
print('=' * 60)

# Update networkconfig.json
print('\nUpdating networkconfig.json...')
with open('networkconfig.json', 'r') as f:
    config = json.load(f)

config['ethereum']['contracts']['StateBloater']['address'] = contract_address

with open('networkconfig.json', 'w') as f:
    json.dump(config, f, indent=2)

print('✓ Config updated')
print('\n✅ Ready for baseline testing!\n')

File: deploy_contract.js
const ethers = require('ethers');
const fs = require('fs');

async function deploy() {
    console.log('Deploying StateBloater contract...\n');

    // Connect to Besu with polling disabled (avoid "latest" block issues)
    const provider = new ethers.JsonRpcProvider('http://127.0.0.1:8545', {
        name: 'besu',
        chainId: 1337
    }, {
        staticNetwork: true,  // Don't query network details
        batchMaxCount: 1
    });

    // Setup wallet with private key
    const privateKey = '0x8f2a55949038a9610f502c24114d051185071191bc20b60811a2d7fba4513689';
    const wallet = new ethers.Wallet(privateKey, provider);

    console.log(`✓ Connected to Besu`);
    console.log(`✓ Using deployer: ${wallet.address}`);

    // Load contract
    const artifact = JSON.parse(fs.readFileSync('./StateBloater.json', 'utf8'));

    console.log('\n🚀 Deploying contract...');

    // Create and sign transaction manually
    const factory = new ethers.ContractFactory(artifact.abi, artifact.bytecode, wallet);

    // Get deploy transaction
    const deployTx = await factory.getDeployTransaction();

    // Manual transaction parameters (ensure data is set explicitly)
    const tx = {
        data: deployTx.data,
        gasLimit: 8000000,
        gasPrice: 0,
        chainId: 1337
    };

    console.log('  Sending deploy transaction...');
    const sentTx = await wallet.sendTransaction(tx);

    console.log(`✓ Transaction hash: ${sentTx.hash}`);
    console.log('  Waiting for confirmation...');

    const receipt = await sentTx.wait();
    const address = receipt.contractAddress;

    console.log('\n' + '='.repeat(60));
    console.log('✅ DEPLOYMENT SUCCESSFUL');
    console.log('='.repeat(60));
    console.log(`Contract Address: ${address}`);
    console.log('='.repeat(60));

    // Update networkconfig.json
    console.log('\nUpdating networkconfig.json...');
    const config = JSON.parse(fs.readFileSync('./networkconfig.json', 'utf8'));
    config.ethereum.contracts.StateBloater.address = address;
    fs.writeFileSync('./networkconfig.json', JSON.stringify(config, null, 2));
    console.log('✓ Config updated');

    console.log('\n✅ Ready for baseline testing!\n');
}

deploy().catch(err => {
    console.error('\n❌ Deployment failed:');
    console.error(err.message || err);
    process.exit(1);
});
